File: simpleciscotui/cisco.py
"""Cisco IOS device access for SimpleCiscoTUI.

Connections are driven through the **system ``ssh`` binary**, not a Python SSH
library. That means the app inherits everything in your ``~/.ssh/config`` (keys,
host aliases, and the legacy crypto old IOS needs) and works wherever plain
``ssh <host>`` already works — which sidesteps the keyboard-interactive and
legacy-algorithm problems Paramiko hits on older switches.

The pure parsing/command helpers (:func:`parse_interface_brief`,
:func:`parse_interface_acls`, :func:`build_apply_commands`, …) carry no I/O and
are unit-tested against captured ``show`` output. Everything that touches the
wire lives on :class:`CiscoSession`.
"""

from __future__ import annotations

import os
import re
import shutil
from dataclasses import dataclass, field
from typing import Optional

try:  # pragma: no cover - import guard
    import pexpect
except Exception:  # pragma: no cover
    pexpect = None  # type: ignore[assignment]

DIRECTIONS = ("in", "out")


@dataclass
class Interface:
    """One row of ``show ip interface brief``."""

    name: str
    ip: str
    status: str
    protocol: str

    @property
    def is_up(self) -> bool:
        return self.status.lower() == "up" and self.protocol.lower() == "up"


@dataclass
class InterfaceAcls:
    """ACLs bound to an interface, keyed by direction."""

    inbound: Optional[str] = None
    outbound: Optional[str] = None

    def get(self, direction: str) -> Optional[str]:
        return self.inbound if direction == "in" else self.outbound

    def bindings(self) -> list[tuple[str, str]]:
        """Return ``(direction, acl_name)`` pairs that are actually set."""
        out: list[tuple[str, str]] = []
        if self.inbound:
            out.append(("in", self.inbound))
        if self.outbound:
            out.append(("out", self.outbound))
        return out


@dataclass
class CiscoCredentials:
    """How to reach one device via the system ``ssh`` client.

    With no ``password`` the app uses key / agent auth and never blocks on a
    prompt (``BatchMode=yes``) — exactly like ``ssh <host>``. ``key_file`` and
    ``legacy_ssh`` map onto ``ssh -i`` / ``ssh -o`` flags, so a device works
    even without a matching ``~/.ssh/config`` block.
    """

    host: str
    username: str = ""
    password: str = ""  # only usable if `sshpass` is installed; key auth preferred
    secret: str = ""  # enable secret; sent after login when set
    port: int = 22
    key_file: Optional[str] = None  # explicit private key path
    legacy_ssh: bool = False  # add the crypto old IOS (e.g. 2960G) requires

    @property
    def uses_password(self) -> bool:
        return bool(self.password)

    def target(self) -> str:
        return f"{self.username}@{self.host}" if self.username else self.host

    def ssh_options(self, conn_timeout: int = 25) -> list[str]:
        """The ``ssh`` flags (without the target host) for this device."""
        opts = ["-o", f"ConnectTimeout={conn_timeout}"]
        if not self.password:
            # Key / agent auth only — never hang waiting for a password prompt
            # the TUI can't display.
            opts += ["-o", "BatchMode=yes"]
        if self.port and self.port != 22:
            opts += ["-p", str(self.port)]
        if self.key_file:
            opts += ["-o", "IdentitiesOnly=yes", "-i", os.path.expanduser(self.key_file)]
        if self.legacy_ssh:
            opts += [
                "-o",
                "KexAlgorithms=+diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha1",
                "-o", "HostKeyAlgorithms=+ssh-rsa",
                "-o", "PubkeyAcceptedAlgorithms=+ssh-rsa",
                "-o", "Ciphers=+aes128-cbc,aes192-cbc,aes256-cbc,3des-cbc",
            ]
        return opts


# --------------------------------------------------------------------------- #
# Pure parsing / command helpers (no I/O)
# --------------------------------------------------------------------------- #

_BRIEF_HEADER = re.compile(r"^\s*Interface\s+IP-Address", re.IGNORECASE)


def parse_interface_brief(text: str) -> list[Interface]:
    """Parse ``show ip interface brief`` into :class:`Interface` rows."""
    interfaces: list[Interface] = []
    for line in text.splitlines():
        if not line.strip() or _BRIEF_HEADER.match(line):
            continue
        parts = line.split()
        if len(parts) < 6:
            # Skips blank lines, echoed commands, and the device prompt.
            continue
        name, ip = parts[0], parts[1]
        status, protocol = parts[-2], parts[-1]
        if not re.match(r"^[A-Za-z]", name):
            continue
        interfaces.append(Interface(name=name, ip=ip, status=status, protocol=protocol))
    return interfaces


def parse_interface_acls(text: str) -> InterfaceAcls:
    """Parse ``ip access-group <name> <in|out>`` lines from interface config."""
    acls = InterfaceAcls()
    for match in re.finditer(r"ip access-group\s+(\S+)\s+(in|out)", text, re.IGNORECASE):
        name, direction = match.group(1), match.group(2).lower()
        if direction == "in":
            acls.inbound = name
        else:
            acls.outbound = name
    return acls


def normalize_direction(direction: str) -> str:
    direction = direction.strip().lower()
    if direction not in DIRECTIONS:
        raise ValueError(f"direction must be one of {DIRECTIONS}, got {direction!r}")
    return direction


def build_apply_commands(interface: str, acl: str, direction: str) -> list[str]:
    """Config-mode commands to bind ``acl`` to ``interface``."""
    direction = normalize_direction(direction)
    return [f"interface {interface}", f"ip access-group {acl} {direction}"]


def build_remove_commands(interface: str, acl: str, direction: str) -> list[str]:
    """Config-mode commands to unbind ``acl`` from ``interface``."""
    direction = normalize_direction(direction)
    return [f"interface {interface}", f"no ip access-group {acl} {direction}"]


#: Matches an IOS exec/config prompt at the end of the buffer, e.g.
#: ``c2960r1#``, ``c2960r1>``, ``c2960r1(config)#``, ``c2960r1(config-if)#``.
PROMPT = r"[A-Za-z0-9._\-]+(?:\([^)]+\))?[>#] ?$"


def _clean_output(text: str) -> str:
    """Normalise pexpect output: strip CRs and the leading echoed command."""
    return text.replace("\r", "")


# --------------------------------------------------------------------------- #
# Live session (drives the system ssh client through a PTY via pexpect)
# --------------------------------------------------------------------------- #


@dataclass
class CiscoSession:
    """An interactive IOS session over the system ``ssh`` client.

    ``ssh`` is spawned once under a pseudo-terminal (via ``pexpect``) and kept
    open; commands are sent one at a time and read back up to the device prompt.
    This is what lets old IOS — which dribbles its interactive parser one line
    at a time — work reliably, unlike piping a whole script at once.
    """

    credentials: CiscoCredentials
    conn_timeout: int = 25
    command_timeout: int = 60
    _child: object = field(default=None, repr=False)
    _connected: bool = field(default=False, repr=False)

    @property
    def connected(self) -> bool:
        return self._connected

    def _ssh_argv(self) -> list[str]:
        creds = self.credentials
        argv = ["ssh", *creds.ssh_options(self.conn_timeout), creds.target()]
        if creds.password and shutil.which("sshpass"):
            argv = ["sshpass", "-p", creds.password, *argv]
        return argv

    # -- lifecycle -------------------------------------------------------- #

    def connect(self) -> None:
        if pexpect is None:  # pragma: no cover
            raise RuntimeError("pexpect is not installed; run `pip install -r requirements.txt`")
        if shutil.which("ssh") is None:  # pragma: no cover
            raise RuntimeError("the `ssh` client was not found on PATH")
        key_file = self.credentials.key_file
        if key_file and not os.path.isfile(os.path.expanduser(key_file)):
            raise FileNotFoundError(f"SSH key not found: {key_file}")

        argv = self._ssh_argv()
        child = pexpect.spawn(argv[0], argv[1:], timeout=self.conn_timeout, encoding="utf-8")
        try:
            child.expect(PROMPT)
        except Exception as exc:  # pexpect EOF/TIMEOUT → auth or connect failure
            detail = _clean_output((child.before or "")).strip().splitlines()
            child.close(force=True)
            msg = detail[-1] if detail else "ssh connection failed"
            raise RuntimeError(msg) from exc
        self._child = child

        # Privileged mode + disable paging.
        if self.credentials.secret and (child.after or "").rstrip().endswith(">"):
            child.sendline("enable")
            child.expect([r"[Pp]assword:", PROMPT])
            if "assword" in (child.after or ""):
                child.sendline(self.credentials.secret)
                child.expect(PROMPT)
        child.sendline("terminal length 0")
        child.expect(PROMPT)
        self._connected = True

    def disconnect(self) -> None:
        child = self._child
        if child is not None:
            try:  # best effort graceful logout
                child.sendline("exit")
                child.expect(pexpect.EOF, timeout=5)
            except Exception:  # noqa: BLE001
                pass
            finally:
                child.close(force=True)
                self._child = None
        self._connected = False

    # -- read ------------------------------------------------------------- #

    def list_interfaces(self) -> list[Interface]:
        return parse_interface_brief(self._run("show ip interface brief"))

    def interface_acls(self, interface: str) -> InterfaceAcls:
        return parse_interface_acls(self._run(f"show running-config interface {interface}"))

    # -- write ------------------------------------------------------------ #

    def apply_acl(self, interface: str, acl: str, direction: str) -> str:
        cmds = ["configure terminal", *build_apply_commands(interface, acl, direction), "end"]
        return self._run_many(cmds)

    def remove_acl(self, interface: str, acl: str, direction: str) -> str:
        cmds = ["configure terminal", *build_remove_commands(interface, acl, direction), "end"]
        return self._run_many(cmds)

    def save_config(self) -> str:
        """``write memory`` — copy running-config to startup-config (no prompt)."""
        return self._run("write memory")

    # -- internals -------------------------------------------------------- #

    def _run(self, command: str) -> str:
        """Send one command, return the device output (echo line stripped)."""
        if self._child is None:
            raise RuntimeError("not connected")
        child = self._child
        child.sendline(command)
        try:
            child.expect(PROMPT, timeout=self.command_timeout)
        except Exception as exc:  # noqa: BLE001
            raise RuntimeError(
                f"timed out waiting for prompt after {command!r}"
            ) from exc
        out = _clean_output(child.before or "")
        # Drop the first line (the echoed command itself).
        return out.split("\n", 1)[1] if "\n" in out else ""

    def _run_many(self, commands: list[str]) -> str:
        return "\n".join(self._run(cmd) for cmd in commands).strip()

File: internal/cisco/cisco.go
// Package cisco provides Cisco IOS device access for SimpleCiscoTUI.
//
// Connections are driven through the system ssh binary, not a Go SSH
// library. That means the app inherits everything in ~/.ssh/config (keys,
// host aliases, and the legacy crypto old IOS needs) and works wherever
// plain `ssh <host>` already works.
//
// The pure parsing/command helpers (ParseInterfaceBrief, ParseInterfaceAcls,
// BuildApplyCommands, ...) carry no I/O and are unit-tested against captured
// `show` output. Everything that touches the wire lives in session.go.
package cisco

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Directions are the two ACL bind directions IOS understands.
var Directions = []string{"in", "out"}

// Interface is one row of `show ip interface brief`.
type Interface struct {
	Name     string
	IP       string
	Status   string
	Protocol string
}

// IsUp reports whether both status and protocol are "up".
func (i Interface) IsUp() bool {
	return strings.EqualFold(i.Status, "up") && strings.EqualFold(i.Protocol, "up")
}

// InterfaceAcls holds the ACLs bound to an interface, keyed by direction.
type InterfaceAcls struct {
	Inbound  string
	Outbound string
}

// Get returns the ACL bound in the given direction ("in" or "out").
func (a InterfaceAcls) Get(direction string) string {
	if direction == "in" {
		return a.Inbound
	}
	return a.Outbound
}

// Binding is a (direction, aclName) pair.
type Binding struct {
	Direction string
	ACL       string
}

// Bindings returns the (direction, acl) pairs that are actually set.
func (a InterfaceAcls) Bindings() []Binding {
	var out []Binding
	if a.Inbound != "" {
		out = append(out, Binding{"in", a.Inbound})
	}
	if a.Outbound != "" {
		out = append(out, Binding{"out", a.Outbound})
	}
	return out
}

// Credentials describes how to reach one device via the system ssh client.
//
// With no Password the app uses key/agent auth and never blocks on a prompt
// (BatchMode=yes) -- exactly like `ssh <host>`. KeyFile and LegacySSH map
// onto `ssh -i` / `ssh -o` flags, so a device works even without a matching
// ~/.ssh/config block.
type Credentials struct {
	Host      string
	Username  string
	Password  string // only usable if `sshpass` is installed; key auth preferred
	Secret    string // enable secret; sent after login when set
	Port      int
	KeyFile   string // explicit private key path
	LegacySSH bool   // add the crypto old IOS (e.g. 2960G) requires
}

// UsesPassword reports whether these credentials carry a password.
func (c Credentials) UsesPassword() bool {
	return c.Password != ""
}

// Target returns "user@host", or just "host" with no username.
func (c Credentials) Target() string {
	if c.Username != "" {
		return c.Username + "@" + c.Host
	}
	return c.Host
}

// SSHOptions returns the `ssh` flags (without the target host) for this device.
func (c Credentials) SSHOptions(connTimeout time.Duration) []string {
	secs := int(connTimeout / time.Second)
	opts := []string{"-o", fmt.Sprintf("ConnectTimeout=%d", secs)}
	if c.Password == "" {
		// Key / agent auth only -- never hang waiting for a password prompt
		// the TUI can't display.
		opts = append(opts, "-o", "BatchMode=yes")
	}
	if c.Port != 0 && c.Port != 22 {
		opts = append(opts, "-p", strconv.Itoa(c.Port))
	}
	if c.KeyFile != "" {
		opts = append(opts, "-o", "IdentitiesOnly=yes", "-i", expandHome(c.KeyFile))
	}
	if c.LegacySSH {
		opts = append(opts,
			"-o", "KexAlgorithms=+diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha1",
			"-o", "HostKeyAlgorithms=+ssh-rsa",
			"-o", "PubkeyAcceptedAlgorithms=+ssh-rsa",
			"-o", "Ciphers=+aes128-cbc,aes192-cbc,aes256-cbc,3des-cbc",
		)
	}
	return opts
}

func expandHome(path string) string {
	if path == "~" {
		home, err := os.UserHomeDir()
		if err != nil {
			return path
		}
		return home
	}
	if strings.HasPrefix(path, "~/") {
		home, err := os.UserHomeDir()
		if err != nil {
			return path
		}
		return filepath.Join(home, path[2:])
	}
	return path
}

// --------------------------------------------------------------------------
// Pure parsing / command helpers (no I/O)
// --------------------------------------------------------------------------

var briefHeaderRe = regexp.MustCompile(`(?i)^\s*Interface\s+IP-Address`)

// ParseInterfaceBrief parses `show ip interface brief` into Interface rows.
func ParseInterfaceBrief(text string) []Interface {
	var interfaces []Interface
	for _, line := range strings.Split(text, "\n") {
		if strings.TrimSpace(line) == "" || briefHeaderRe.MatchString(line) {
			continue
		}
		parts := strings.Fields(line)
		if len(parts) < 6 {
			// Skips blank lines, echoed commands, and the device prompt.
			continue
		}
		name, ip := parts[0], parts[1]
		status, protocol := parts[len(parts)-2], parts[len(parts)-1]
		if name == "" || !isAsciiLetter(rune(name[0])) {
			continue
		}
		interfaces = append(interfaces, Interface{Name: name, IP: ip, Status: status, Protocol: protocol})
	}
	return interfaces
}

func isAsciiLetter(r rune) bool {
	return (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z')
}

var aclBindingRe = regexp.MustCompile(`(?i)ip access-group\s+(\S+)\s+(in|out)`)

// ParseInterfaceAcls parses `ip access-group <name> <in|out>` lines from
// interface config.
func ParseInterfaceAcls(text string) InterfaceAcls {
	var acls InterfaceAcls
	for _, m := range aclBindingRe.FindAllStringSubmatch(text, -1) {
		name, direction := m[1], strings.ToLower(m[2])
		if direction == "in" {
			acls.Inbound = name
		} else {
			acls.Outbound = name
		}
	}
	return acls
}

// NormalizeDirection lower-cases and validates a direction string.
func NormalizeDirection(direction string) (string, error) {
	direction = strings.ToLower(strings.TrimSpace(direction))
	if direction != "in" && direction != "out" {
		return "", fmt.Errorf("direction must be one of %v, got %q", Directions, direction)
	}
	return direction, nil
}

// BuildApplyCommands returns the config-mode commands to bind acl to interface.
func BuildApplyCommands(iface, acl, direction string) ([]string, error) {
	direction, err := NormalizeDirection(direction)
	if err != nil {
		return nil, err
	}
	return []string{
		"interface " + iface,
		fmt.Sprintf("ip access-group %s %s", acl, direction),
	}, nil
}

// BuildRemoveCommands returns the config-mode commands to unbind acl from interface.
func BuildRemoveCommands(iface, acl, direction string) ([]string, error) {
	direction, err := NormalizeDirection(direction)
	if err != nil {
		return nil, err
	}
	return []string{
		"interface " + iface,
		fmt.Sprintf("no ip access-group %s %s", acl, direction),
	}, nil
}

// PromptRe matches an IOS exec/config prompt at the end of the buffer, e.g.
// "c2960r1#", "c2960r1>", "c2960r1(config)#", "c2960r1(config-if)#".
var PromptRe = regexp.MustCompile(`[A-Za-z0-9._\-]+(?:\([^)]+\))?[>#] ?$`)

// PasswordPromptRe matches ssh/enable password prompts.
var PasswordPromptRe = regexp.MustCompile(`[Pp]assword:`)

// cleanOutput normalises raw pty output: strip CRs.
func cleanOutput(text string) string {
	return strings.ReplaceAll(text, "\r", "")
}
